### PR TITLE
add numbered aliases [#154548835]

### PR DESCRIPTION
**Pull request by uraniumanchor**
# Contributing to the Donation Tracker

- [X] I've added tests or modified existing tests for the change.
- [X] I've humanly end-to-end tested the change by running an instance of the tracker.

### Issue from Pivotal Tracker

https://www.pivotaltracker.com/story/show/154548835

### Description of the Change

When people donate using a different email than before (as decided by their PayPal contact email), but try to use the same alias, it puts a de-duping number on the end of it. This change makes that its own column (and moves the uniqueness constraint on the pair of them) so that it's easier to discern what people put in deliberately vs what the system had to add. Functions similarly to BNet or Discord. 

Note that this doesn't stop people from putting hash marks in their names, the number is a separate field and easily distinguishable via the api.

### Verification Process

There were two complicated parts to this.

The backfill got tackled first, I made some test donations and then ran the migration and everything seemed to work correctly.

The other part was the API and the donation processing pages. I realized there were no tests for a `donation` search so I added some. The processing pages don't have any explicit tests (since they're old crusty jQuery) so I tested them thoroughly by hand with both one-step and two-step screening, and with various levels of access (analogous to what we use at the events) and everything seemed to work correctly. One semi-related change I made to the behavior is that the processing pages no longer try to make links to the Donor editor if the user can't actually edit donors (they would have 404'ed anyway, but might as well not confuse people).

Lastly, since this does affect IPN processing, I did a sandbox donation and made sure it didn't explode.

### CODE REVIEW

File: tracker/migrations/0010_add_alias_no.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('tracker', '0009_donation_request_fields_blank'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='donor',
+            name='alias_no',
```
Comment (yuna9) [inline tracker/migrations/0010_add_alias_no.py:17]: think num is clearer than no

File: tracker/models/donation.py [hunk 2]
```diff
@@ -347,15 +358,29 @@ class Meta:
...
         ordering = ['lastname', 'firstname', 'email']
+        unique_together = [('alias', 'alias_no')]
 
-    def clean(self):
+    def save(self, *args, **kwargs):
         # an empty value means a null value
         if not self.alias:
             self.alias = None
+            self.alias_no = None
+        elif not self.alias_no:
```
Comment (yuna9) [inline tracker/models/donation.py:368]: am i missing the part of this pr where the alias_no is redrawn if the alias changes and causes a conflict?

File: tracker/paypalutil.py [hunk 3]
```diff
@@ -130,15 +129,8 @@ def initialize_paypal_donation(ipnObj):
...
-            currentAlias = donation.requestedalias
-            while not foundAResult:
-                results = Donor.objects.filter(alias__iexact=currentAlias)
-                if results.exists():
-                    currentAlias = donation.requestedalias + str(random.getrandbits(8))
-                else:
-                    foundAResult = True
-            donor.alias = currentAlias
+            donor.alias = donation.requestedalias
+            donor.alias_no = None  # will get filled in by the donor save
```
Comment (yuna9) [inline tracker/paypalutil.py:133]: is it this?

**Review by yuna9 — APPROVED** [review]
lgtm, think a look from faulty can't hurt

File: tracker/paypalutil.py (hunk 3 above)
Comment (uraniumanchor) [reply tracker/paypalutil.py:133]: Yes, this is the only place currently where `requestedalias` actually does anything.

File: tracker/migrations/0010_add_alias_no.py (hunk 1 above)
Comment (yuna9) [reply tracker/migrations/0010_add_alias_no.py:17]: any comment on this?

Comment (uraniumanchor) [reply tracker/migrations/0010_add_alias_no.py:17]: I'll change it before I merge, yeah. This code is originally from like 2018 so I don't remember why I picked `no` and I can't think of a good reason to keep it.

**Pull request merged**